Source/DafnyLanguageServer.Test/Lookup/HoverVerificationTest.cs

```diff
@@ -0,0 +1,172 @@
+*assertion might not hold*"
+      );
+      await AssertHoverMatches(documentItem, (2, 12),
+        @"assertion of [batch](???) #???/2 checked in ???ms with ??? resource count:  
+*assertion always holds*"
+      );
+    }
+
+    [TestMethod, Timeout(MaxTestExecutionTimeMs)]
+    public async Task MeaningfulMessageWhenMethodWithoutAssert() {
+      var documentItem = await GetDocumentItem(@"
+method f(x: int) {
+  print x;
+}", "testfile.dfy", CompilationStatus.VerificationSucceeded);
+      await Task.Delay(100); // Just time for the diagnostics to be updated
+      await AssertHoverMatches(documentItem, (0, 7),
+        @"**f** metrics:
+
+No assertion to check."
+      );
+      await AssertHoverMatches(documentItem, (0, 10),
+        "```dafny\nx: int\n```");
+    }
+
+    private async Task<TextDocumentItem> GetDocumentItem(string source, string filename, CompilationStatus expectedStatus) {
+      source = source.TrimStart();
+      var documentItem = CreateTestDocument(source, filename);
+      await client.OpenDocumentAndWaitAsync(documentItem, CancellationToken);
+      var lastStatus = await WaitUntilDafnyFinishes(documentItem);
+      Assert.AreEqual(expectedStatus, lastStatus);
+      return documentItem;
+    }
+
+    private async Task AssertHoverMatches(TextDocumentItem documentItem, Position hoverPosition, string expected) {
+      var hover = await RequestHover(documentItem, hoverPosition);
+      Assert.IsNotNull(hover);
+      var markup = hover.Contents.MarkupContent;
+      Assert.IsNotNull(markup);
+      Assert.AreEqual(MarkupKind.Markdown, markup.Kind);
+      AssertMatchRegex(expected, markup.Value);
+    }
+
+    private void AssertMatchRegex(string expected, string value) {
+      var regexExpected = Regex.Escape(expected).Replace(@"\?\?\?", ".*");
+      var matched = new Regex(regexExpected).Match(value).Success;
+      if (!matched) {
+        // A simple helper to determine what portion of the regex did not match
+        var helper = "";
+        foreach (var chunk in expected.Split("???")) {
+          if (!value.Contains(chunk)) {
+            helper += $"\nThe result string did not contain '{chunk}'";
+          }
+        }
+        Assert.IsTrue(false, "{0} did not match {1}." + helper, value, regexExpected);
+      }
+    }
+
+    private async Task<CompilationStatus> WaitUntilDafnyFinishes(TextDocumentItem documentItem) {
+      CompilationStatusParams lastResult;
+      do {
+        lastResult = await notificationReceiver.AwaitNextNotificationAsync(CancellationToken);
+        Assert.AreEqual(documentItem.Uri, lastResult.Uri);
+        Assert.AreEqual(documentItem.Version, lastResult.Version);
+      } while (lastResult.Status != CompilationStatus.VerificationFailed &&
+               lastResult.Status != CompilationStatus.VerificationSucceeded &&
+               lastResult.Status != CompilationStatus.ParsingFailed &&
+               lastResult.Status != CompilationStatus.ResolutionFailed);
```

Comment:
> Would it be worth having a separate definition for "done" that encapsulates these? It would make the code a little less messy, and would make it clear what the meaning of that big conjunction is.

Refactored, thanks for pointing this out.